File: clients/nodejs/test/stress.js
const LightKVClient = require('../src/client');

const NUM_THREADS = 10;
const OPS_PER_THREAD = 1000;

async function worker(threadId, opsCount) {
    const client = new LightKVClient({ host: '127.0.0.1', port: 16379 });
    await client.connect();

    let success = 0;
    let failed = 0;
    const latencies = [];

    for (let i = 0; i < opsCount; i++) {
        const key = `stress_${threadId}_${i}`;
        const value = `value_${i}`;

        const start = Date.now();
        try {
            await client.set(key, value);
            await client.get(key);
            await client.delete(key);
            success += 3;
        } catch (e) {
            failed += 3;
        }
        const ms = Date.now() - start;
        latencies.push(ms);
    }

    await client.quit();
    return { success, failed, latencies };
}

async function main() {
    const totalOps = NUM_THREADS * OPS_PER_THREAD * 3;

    console.log('[Node.js Stress Test]');
    console.log(`  Threads: ${NUM_THREADS}`);
    console.log(`  Ops/thread: ${OPS_PER_THREAD} (SET+GET+DELETE)`);
    console.log(`  Total ops: ${totalOps}`);

    const start = Date.now();

    const promises = [];
    for (let i = 0; i < NUM_THREADS; i++) {
        promises.push(worker(i, OPS_PER_THREAD));
    }

    const results = await Promise.all(promises);
    const duration = Date.now() - start;

    let totalSuccess = 0;
    let totalFailed = 0;
    const allLatencies = [];

    for (const r of results) {
        totalSuccess += r.success;
        totalFailed += r.failed;
        allLatencies.push(...r.latencies);
    }

    allLatencies.sort((a, b) => a - b);
    const p50 = allLatencies[Math.floor(allLatencies.length * 0.5)];
    const p99 = allLatencies[Math.floor(allLatencies.length * 0.99)];
    const avg = allLatencies.reduce((a, b) => a + b, 0) / allLatencies.length;
    const opsSec = Math.round(totalSuccess / (duration / 1000));

    console.log('\n  Results:');
    console.log(`    Duration:     ${duration} ms`);
    console.log(`    Success:      ${totalSuccess} ops`);
    console.log(`    Failed:       ${totalFailed} ops`);
    console.log(`    Throughput:   ${opsSec.toLocaleString()} ops/sec`);
    console.log(`    Avg Latency:  ${avg.toFixed(2)} ms`);
    console.log(`    P50 Latency:  ${p50} ms`);
    console.log(`    P99 Latency:  ${p99} ms`);

    console.log('\n[Node.js Stress Test Complete]');
}

main().catch(err => {
    console.error('Error:', err);
    process.exit(1);
});

File: clients/python/tests/stress.py
"""LightKV Python Client Stress Test."""

import time
import sys
import os
import threading
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'lightkv'))

from client import LightKVClient

NUM_THREADS = 10
OPS_PER_THREAD = 1000

results_lock = threading.Lock()
total_success = 0
total_failed = 0
all_latencies = []


def worker(thread_id, ops_count):
    global total_success, total_failed

    client = LightKVClient(host='127.0.0.1', port=16379)
    if not client.connect():
        with results_lock:
            total_failed += ops_count * 3
        return

    success = 0
    failed = 0
    latencies = []

    for i in range(ops_count):
        key = f'stress_{thread_id}_{i}'
        value = f'value_{i}'

        start = time.time()
        try:
            client.set(key, value)
            client.get(key)
            client.delete(key)
            success += 3
        except Exception:
            failed += 3
        ms = (time.time() - start) * 1000
        latencies.append(ms)

    client.quit()

    with results_lock:
        total_success += success
        total_failed += failed
        all_latencies.extend(latencies)


def main():
    total_ops = NUM_THREADS * OPS_PER_THREAD * 3

    print('[Python Stress Test]')
    print(f'  Threads: {NUM_THREADS}')
    print(f'  Ops/thread: {OPS_PER_THREAD} (SET+GET+DELETE)')
    print(f'  Total ops: {total_ops}')

    start = time.time()

    threads = []
    for i in range(NUM_THREADS):
        t = threading.Thread(target=worker, args=(i, OPS_PER_THREAD))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    duration = (time.time() - start) * 1000

    all_latencies.sort()
    p50 = all_latencies[len(all_latencies) // 2]
    p99 = all_latencies[int(len(all_latencies) * 0.99)]
    avg = sum(all_latencies) / len(all_latencies)
    ops_sec = total_success / (duration / 1000)

    print(f'\n  Results:')
    print(f'    Duration:     {duration:.1f} ms')
    print(f'    Success:      {total_success} ops')
    print(f'    Failed:       {total_failed} ops')
    print(f'    Throughput:   {int(ops_sec):,} ops/sec')
    print(f'    Avg Latency:  {avg:.2f} ms')
    print(f'    P50 Latency:  {p50:.2f} ms')
    print(f'    P99 Latency:  {p99:.2f} ms')

    print(f'\n[Python Stress Test Complete]')


if __name__ == '__main__':
    main()
